### PR TITLE
Update DNS document to talk about the new Pod and Endpoints Annotations to set hostname

### PR DESCRIPTION
**Pull request by ArtfulCoder**
…

### CODE REVIEW

File: cluster/addons/dns/README.md [hunk 1]
```diff
@@ -59,6 +59,21 @@ When enabled, pods are assigned a DNS A record in the form of `pod-ip-address.my
 
 For example, a pod with ip `1.2.3.4` in the namespace `default` with a dns name of `cluster.local` would have an entry: `1-2-3-4.default.pod.cluster.local`.
 
+
+####A Records and hostname Based on Pod Annotations - A Beta Feature in Kubernetes v1.3
```
Comment (thockin) [inline cluster/addons/dns/README.md]: v1.2 not v1.3 :)

File: cluster/addons/dns/README.md [hunk 2]
```diff
@@ -59,6 +59,21 @@ When enabled, pods are assigned a DNS A record in the form of `pod-ip-address.my
...
+Currently when a pod is created it's hostname is the Pod's Name value.
```
Comment (thockin) [inline cluster/addons/dns/README.md]: it's 'its' not it's.

Comment (thockin) [inline cluster/addons/dns/README.md]: s/Name/`metadata.name`/

File: cluster/addons/dns/README.md [hunk 3]
```diff
@@ -59,6 +59,21 @@ When enabled, pods are assigned a DNS A record in the form of `pod-ip-address.my
...
+With v1.3, users can specify a Pod Annotation, pod.beta.kubernetes.io/hostname, to specify what the Pod's hostname should be.
```
Comment (thockin) [inline cluster/addons/dns/README.md]: v1.2

Comment (thockin) [inline cluster/addons/dns/README.md]: s/Annotation/annotation/

Comment (thockin) [inline cluster/addons/dns/README.md]: backtick-quote the annotation name

Comment (thockin) [inline cluster/addons/dns/README.md]: add an example:  e.g. `pod.beta.kubernetes.io/hostname: my-pod-name` will result in your Pod seeing `my-pod-name` as its hostname.

File: cluster/addons/dns/README.md [hunk 4]
```diff
@@ -59,6 +59,21 @@ When enabled, pods are assigned a DNS A record in the form of `pod-ip-address.my
...
+If the annotation is specified, the annotation value takes precendence over the Pod's Name, to be the hostname of the pod.
```
Comment (thockin) [inline cluster/addons/dns/README.md]: s/Name/name

File: cluster/addons/dns/README.md [hunk 5]
```diff
@@ -59,6 +59,21 @@ When enabled, pods are assigned a DNS A record in the form of `pod-ip-address.my
...
+
+v1.3 introduces a beta feature where the user can specify a Pod Annotation, pod.beta.kubernetes.io/subdomain, to specify what the Pod's subdomain should be.
```
Comment (thockin) [inline cluster/addons/dns/README.md]: v1.2

Comment (thockin) [inline cluster/addons/dns/README.md]: s/Annotation/annotation/

Comment (thockin) [inline cluster/addons/dns/README.md]: backtick-quote the annotation name

File: cluster/addons/dns/README.md [hunk 6]
```diff
@@ -59,6 +59,21 @@ When enabled, pods are assigned a DNS A record in the form of `pod-ip-address.my
...
+If the annotation is specified, the fully qualified Pod hostname will be <hostname>.<subdomain>.<pod namespace>.svc.<cluster domain>
```
Comment (thockin) [inline cluster/addons/dns/README.md]: example

File: cluster/addons/dns/README.md [hunk 7]
```diff
@@ -59,6 +59,21 @@ When enabled, pods are assigned a DNS A record in the form of `pod-ip-address.my
...
+
+If there exists a headless service in the same namespace as the pod and with the same name as the subdomain, the cluster's KubeDNS Server will also return an A record for the Pod's fully qualified hostname.
```
Comment (thockin) [inline cluster/addons/dns/README.md]: This needs an example.

Given a Pod with the hostname annotation set to "foo" and the subdomain annotation set to "bar", and a headless Service named "bar", the pod will see it's own FQDN as "foo.bar.my-namespace.svc.cluster.local".  DNS will serve an A record at that name, pointing to the Pod's IP.

File: cluster/addons/dns/README.md [hunk 8]
```diff
@@ -59,6 +59,21 @@ When enabled, pods are assigned a DNS A record in the form of `pod-ip-address.my
...
+
+With v1.3, the Endpoints object also has a new Annotation endpoints.beta.kubernetes.io/hostnames-map. Its value is the json representation of map[string(IP)][endpoints.HostRecord]
+For example: '{"10.245.1.6":{HostName: "my-webserver"}}'. If the Endpoints are for a headless service, then A records will be created with the format <hostname>.<subdomain>.<pod namespace>.svc.<cluster domain>
+For the example json, the subdomain would be headless service name, and the hostname would be  "my-webserver" and the A record lookup would return 10.245.1.6
```
Comment (thockin) [inline cluster/addons/dns/README.md]: I would add that this is not generally needed for end-users but is used by the internal service controller to deliver the aforementioned feature.

File: cluster/addons/dns/README.md (hunk 4 above)
Comment (ArtfulCoder) [inline cluster/addons/dns/README.md]: done

File: cluster/addons/dns/README.md (hunk 2 above)
Comment (ArtfulCoder) [inline cluster/addons/dns/README.md]: done

File: cluster/addons/dns/README.md (hunk 1 above)
Comment (ArtfulCoder) [inline cluster/addons/dns/README.md]: done

File: cluster/addons/dns/README.md (hunk 5 above)
Comment (ArtfulCoder) [inline cluster/addons/dns/README.md]: done

File: cluster/addons/dns/README.md (hunk 8 above)
Comment (ArtfulCoder) [inline cluster/addons/dns/README.md]: done

File: cluster/addons/dns/README.md (hunk 7 above)
Comment (ArtfulCoder) [inline cluster/addons/dns/README.md]: done

File: cluster/addons/dns/README.md (hunk 3 above)
Comment (ArtfulCoder) [inline cluster/addons/dns/README.md]: done

File: cluster/addons/dns/README.md (hunk 6 above)
Comment (ArtfulCoder) [inline cluster/addons/dns/README.md]: done

**Comment by ArtfulCoder** [comment]
PTAL

**Comment by bgrant0607** [comment]
DNS needs a doc in the doc repo.

**Comment by bgrant0607** [comment]
e2e flaked, but this is a doc change. Manually merging. @k8s-oncall

**Comment by eparis** [comment]
This PR is included in #22874 which is slated to be included in the release-1.2 branch.
Please verify that the cherry-pick in that PR looks correct.

**Pull request merged**